src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCompileActionBuilder.java

```diff
@@ -589,168 +563,42 @@ static CustomCommandLine createCommandLineFromToolchains(
-    // - If the source root is under the output directory (itself determined in a hacky way and
-    // relying on the fact that the output roots of all repositories are under the same directory
-    // under the exec root), we check whether the .proto file is under it. If so, we have a match.
-    // - Otherwise, we check whether the .proto file is either under that source directory or under
-    // bin or genfiles by prefix-matching its root-relative path.
-    if (sourceRootPath.segmentCount() > 0 && sourceRootPath.getSegment(0).equals(outputDirectory)) {
-      if (proto.getExecPath().startsWith(sourceRootPath)) {
-        return proto.getExecPath().relativeTo(sourceRootPath).getPathString();
-      }
-    } else {
-      PathFragment prefix =
-          siblingRepositoryLayout
-              ? LabelConstants.EXPERIMENTAL_EXTERNAL_PATH_PREFIX
-              : LabelConstants.EXTERNAL_PATH_PREFIX;
-      if (proto.getRootRelativePath().startsWith(sourceRootPath)) {
-        return proto.getRootRelativePath().relativeTo(sourceRootPath).getPathString();
-      } else if (proto.getExecPath().startsWith(prefix)
-          && proto.getExecPath().startsWith(sourceRootPath)) {
-        return proto.getExecPath().relativeTo(sourceRootPath).getPathString();
-      }
-    }
-
-    return null;
-  }
-
   @AutoCodec @AutoCodec.VisibleForSerialization
   static final CommandLineItem.MapFn<String> EXPAND_TRANSITIVE_PROTO_PATH_FLAGS =
-      (flag, args) -> {
-        if (!flag.equals(".")) {
-          args.accept("--proto_path=" + flag);
-        }
-      };
-
-
-  @AutoCodec
-  @AutoCodec.VisibleForSerialization
-  static final class ExpandImportArgsFn implements CapturingMapFn<Artifact> {
-    private final String outputDirectory;
-    private final NestedSet<String> directProtoSourceRoots;
-    private final boolean siblingRepositoryLayout;
-
-    public ExpandImportArgsFn(
-        String outputDirectory,
-        NestedSet<String> directProtoSourceRoots,
-        boolean siblingRepositoryLayout) {
-      this.outputDirectory = outputDirectory;
-      this.directProtoSourceRoots = directProtoSourceRoots;
-      this.siblingRepositoryLayout = siblingRepositoryLayout;
-    }
-
-    /**
-     * Generates up to two import flags for each artifact: one for full path (only relative to the
-     * repository root) and one for the path relative to the proto source root (if one exists
-     * corresponding to the artifact).
-     */
-    @Override
-    public void expandToCommandLine(Artifact proto, Consumer<String> args) {
-      for (String directProtoSourceRoot : directProtoSourceRoots.toList()) {
-        PathFragment sourceRootPath = PathFragment.create(directProtoSourceRoot);
-        String arg =
-            guessProtoPathUnderRoot(
-                outputDirectory, sourceRootPath, proto, siblingRepositoryLayout);
-        if (arg != null) {
-          args.accept("-I" + arg + "=" + proto.getExecPathString());
```

Comment:
Why was this removed? This is important because this is how `protoc` learns the import map -> file path map.

Reply:
There are two ways to use `protoc`:
  - The "explicit import map"-approach we're using today, and
  - the "C++ #include like"-approach we'll use when this patch lands (i.e. `--proto_path=root1 --proto_path=root2`).

Both can't really deal with ambiguous import paths. `protoc` will happily accept `-Ifoo.proto=/path/to/a.proto -Ifoo.proto=/path/to/b.proto`, but which `foo.proto` it resolves to is, AFAICT, undefined. With `--proto_path`, `protoc` just uses the first matching file it can find.

If we continue with the "import map"-approach, we'd have to provide access to the transitive closure of import path -> file, which sounds like it requires a lot of memory (or still guess them, which would make this whole patch obsolete).

Given the above, and there's no real benefit in using the "explicit import map"-approach vs. "C++ #include like"-approach (`protoc` is negligibly slower with the new approach, but computing the command-line is slightly faster), I decided to drop the import map and let `protoc` handle it.

Reply:
Problem is, you can't really use the `--proto_path` approach because that would result in ambiguous imports (at least in the naive case), 

E.g. if you have both `//a:b.proto' and '@repo//c:d.proto` in your transitive closure, you would need to have both `.` and `external/repo` on your proto path, therefore, `import external/repo/c/d.proto` would work.

The workaround for that would be to create a little import directory for every `proto_library`, but that causes O(n^2) import lookup behavior and is thus unacceptable (I know, I tried because that's the cleanest approach)

The import map does not necessarily require a lot of RAM because the command line is constructed lazily (that's what all those mapping functions are for) and all you need is a description of how the command line is built, which is O(1) number of extra objects.

My game plan was to explicitly store the import path in `ProtoInfo` (i.e. no more `null` values in that `Pair`) -- that's also more RAM, but it's at least only O(n) in the number of .proto files. A bigger hammer would be to store a `NestedSet` of `(import path, [proto file])`, but that's not possible to do without changing the Starlark API.